### PR TITLE
Remove Blocking on GENERIC Pool in GET Snapshots

### LINKED ISSUE #69099: Deadlock in get-snapshots API
**Elasticsearch version** (`bin/elasticsearch --version`): 7.11.0

**Plugins installed**: Cloud

**JVM version** (`java -version`): Cloud

**OS version** (`uname -a` if on a Unix-like system): Cloud

**Description of the problem including expected versus actual behavior**:

A cluster was observed with all generic threads waiting here:

```
"elasticsearch[instance-0000000071][generic][T#1]" #33 daemon prio=5 os_prio=0 cpu=5894.11ms elapsed=559821.30s tid=0x00007f0af0746950 nid=0x1dc waiting on condition  [0x00007f0aef5f8000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@15.0.1/Native Method)
	- parking to wait for  <0x00000006d6607c20> (a org.elasticsearch.common.util.concurrent.BaseFuture$Sync)
	at java.util.concurrent.locks.LockSupport.park(java.base@15.0.1/LockSupport.java:211)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@15.0.1/AbstractQueuedSynchronizer.java:714)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(java.base@15.0.1/AbstractQueuedSynchronizer.java:1046)
	at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:248)
	at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:76)
	at org.elasticsearch.common.util.concurrent.FutureUtils.get(FutureUtils.java:45)
	at org.elasticsearch.action.support.AdapterActionFuture.actionGet(AdapterActionFuture.java:26)
	at org.elasticsearch.action.support.PlainActionFuture.get(PlainActionFuture.java:22)
	at org.elasticsearch.action.admin.cluster.snapshots.get.TransportGetSnapshotsAction.masterOperation(TransportGetSnapshotsAction.java:90)
	at org.elasticsearch.action.admin.cluster.snapshots.get.TransportGetSnapshotsAction.masterOperation(TransportGetSnapshotsAction.java:54)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction.masterOperation(TransportMasterNodeAction.java:88)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.lambda$doStart$3(TransportMasterNodeAction.java:155)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$$Lambda$5714/0x0000000801b184c8.accept(Unknown Source)
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:732)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@15.0.1/ThreadPoolExecutor.java:1130)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@15.0.1/ThreadPoolExecutor.java:630)
	at java.lang.Thread.run(java.base@15.0.1/Thread.java:832)
```

These threads are all waiting to retrieve the repository data, which itself needs generic threads, but there are no more generic threads available.

**Steps to reproduce**:

I have not tried, but I guess that you could reproduce this by hitting the get-snapshots API hard enough, especially if there are lots of old snapshots in the repository so retrieving the `RepositoryData` happens every time and is slow enough.

I suspect this was introduced in https://github.com/elastic/elasticsearch/commit/1cde4a6364ce26ab38c99d2bc08aa4a1d4ce3be6, so might affect all 7.x.y for x ≥ 6, but apparently not 8.z.

### PR DESCRIPTION
**Pull request by original-brownbear**
This aligns the way get snapshots works in `7.x` with how it works
in `master` exactly (except for masters ability to load snapshots across repos)
and thus fixes the deadlock in #69099.

No need to review the logic in detail here, it's the exact same logic as in master now so it probably rather makes sense comparing with master than looking at this in isolation.

Closes #69099

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1  (snapshot test failure but unrelated and known to the features team)

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java [hunk 1]
```diff
@@ -170,16 +181,17 @@ protected void masterOperation(final GetSnapshotsRequest request, final ClusterS
      *                            if false, they will throw an error
      * @return list of snapshots
      */
-    private List<SnapshotInfo> snapshots(@Nullable SnapshotsInProgress snapshotsInProgress, String repositoryName,
+    private List<SnapshotInfo> snapshots(SnapshotsInProgress snapshotsInProgress, String repositoryName,
                                          List<SnapshotId> snapshotIds, boolean ignoreUnavailable) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:185]: Nit: spurious whitespace deviation from master :)

```suggestion
                                        List<SnapshotId> snapshotIds, boolean ignoreUnavailable) {
```

**Review by DaveCTurner — APPROVED** [review]
LGTM

I opened https://github.com/elastic/elasticsearch/issues/69108 to remind us to do something better with the new response format in 8.0 when we can.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:185]: The version here is actually nicer/more correct :D that's why I deviated. I'd rather fix this in master next time a change to the file is made there :)

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**